### PR TITLE
Ruler: add limit parameter in rulegroup

### LINKED ISSUE #4439: Ruler: add limit for number of series produced by recording rules
**Is your feature request related to a problem? Please describe.**
A change has been made [upstream in prometheus](https://github.com/prometheus/prometheus/pull/9260/files) which allows for a limit to be passed to the alerting/recording rules evaluation, to prevent excessive series being produced.

**Describe the solution you'd like**
We should revendor `prometheus/prometheus` to obtain this functionality and expose this limit as a per-tenant and/or global configuration.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by vardhaman-surana**
**What this PR does / why we need it**:
This PR adds the support for group level **limit** parameter in ruler which was introduced in prometheus by PR: https://github.com/prometheus/prometheus/pull/9260

**Which issue(s) this PR fixes**:
Fixes #4439

**Special notes for your reviewer**:

**Checklist**
- [X] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [X] Documentation added
- [X] Tests updated
- [X] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

**Review by dannykopping — COMMENTED** [review]
I've had a brief scan, and it seems like there are no tests for the actual functionality. Did I miss them?

I was expecting to see a test which applies a limit to a rule which returns _n_ series and returns an error because it has exceeded the limit.

I know these tests are covered in Prometheus, but what I'm missing from this PR is something that gives me confidence that the feature would work as intended.

**Comment by vardhaman-surana** [comment]
> I've had a brief scan, and it seems like there are no tests for the actual functionality. Did I miss them?
> 
> I was expecting to see a test which applies a limit to a rule which returns _n_ series and returns an error because it has exceeded the limit.
> 
> I know these tests are covered in Prometheus, but what I'm missing from this PR is something that gives me confidence that the feature would work as intended.

added a test for the functionality. Please review

**Comment by vardhaman-surana** [comment]
> I've had a brief scan, and it seems like there are no tests for the actual functionality. Did I miss them?
> 
> I was expecting to see a test which applies a limit to a rule which returns _n_ series and returns an error because it has exceeded the limit.
> 
> I know these tests are covered in Prometheus, but what I'm missing from this PR is something that gives me confidence that the feature would work as intended.

added a test for the functionality. Please review @dannykopping

**Review by dannykopping — COMMENTED** [review]
Thank you for adding the test @vardhaman-surana 👍 this looks good on first glance.

I'm thinking: if we limit the number of series and a recording rule or alert exceeds that limit, we should probably expose a metric for this. If you had configured this in your hosted Loki environment, how would you want to notice that these limits have been exceeded, and advise your users accordingly? The API response is good, but I think we need more observability here.

Also: please do not force-push changes; we squash merge so there's no reason to rewrite history. Force-pushes make incremental reviews difficult to do.

**Comment by vardhaman-surana** [comment]
@dannykopping i am just getting started with prometheus so maybe my understanding of some concepts will not be perfect.
i am thinking we can add one more metric which will be specific to these limit exceeded cases in rule group metrics something similar to 
[EvalFailures](https://github.com/prometheus/prometheus/blob/cd7d0b69a2344d1508d80c68640a0a5381de9cf6/rules/manager.go#L66)
 and whatever we decide i think it will need to go in prometheus repo first. is my understanding correct? 
 (if true can we merge this PR then ?)
i am exploring loki codebase and have not found anything which can be done in loki yet since rule evaluation etc. all happens in prometheus code.... please point me to the files where this change can go in if possible.
 Thanks

**Comment by dannykopping** [comment]
> i am thinking we can add one more metric which will be specific to these limit exceeded cases in rule group metrics something similar to
[EvalFailures](https://github.com/prometheus/prometheus/blob/cd7d0b69a2344d1508d80c68640a0a5381de9cf6/rules/manager.go#L66)
and whatever we decide i think it will need to go in prometheus repo first. is my understanding correct?

You're correct about the metric, but I'm wondering if we do need to modify Prometheus to get this result... What we could do is check the error that is returned by `Eval()` to see if it matches the expected structure, although that's a little flimsy.

OK, I think for now we can move forward with what we have. We log an error if `Eval()` fails which contains the error message returned by Prometheus about the limit being exceeded, which a user could use if they wanted to alert on this.

If you'd like, you could open an issue in Prometheus to expose a metric.

One last thing: I think we should rename `Limit` to be more specific. It's a bit vague right now, and users might misunderstand its purpose. For example `MaxActiveAlerts` and `MaxRecordingRuleSamples` would be more clear. It's a pity one vague name `Limit` is used for both in Prometheus.

**Comment by vardhaman-surana** [comment]
> > i am thinking we can add one more metric which will be specific to these limit exceeded cases in rule group metrics something similar to
> > [EvalFailures](https://github.com/prometheus/prometheus/blob/cd7d0b69a2344d1508d80c68640a0a5381de9cf6/rules/manager.go#L66)
> > and whatever we decide i think it will need to go in prometheus repo first. is my understanding correct?
> 
> You're correct about the metric, but I'm wondering if we do need to modify Prometheus to get this result... What we could do is check the error that is returned by `Eval()` to see if it matches the expected structure, although that's a little flimsy.
> 
> OK, I think for now we can move forward with what we have. We log an error if `Eval()` fails which contains the error message returned by Prometheus about the limit being exceeded, which a user could use if they wanted to alert on this.
> 
> If you'd like, you could open an issue in Prometheus to expose a metric.
> 
> One last thing: I think we should rename `Limit` to be more specific. It's a bit vague right now, and users might misunderstand its purpose. For example `MaxActiveAlerts` and `MaxRecordingRuleSamples` would be more clear. It's a pity one vague name `Limit` is used for both in Prometheus.

cool then should i will remove limit and add these two fields in loki for this PR. And the log i guess is added in prometheus code already  [here](https://github.com/prometheus/prometheus/blob/e278195e3983c966c2a0f42211f62fa8f40c5561/rules/manager.go#L640)

**Comment by vardhaman-surana** [comment]
hey @dannykopping i started working on it and it seems two fields can not be added till those two fields are supported in prometheus.
Since **limit** is configured at rule group level and it controls both the alerting and recording rules inside that rule group. So even if we expose two fields and a user has both kinds of rules in that group and set both the fields we will only be able to set one to the limit.
Should i open an issue in prometheus for support of two fields ?

**Comment by dannykopping** [comment]
> hey @dannykopping i started working on it and it seems two fields can not be added till those two fields are supported in prometheus. Since **limit** is configured at rule group level and it controls both the alerting and recording rules inside that rule group. So even if we expose two fields and a user has both kinds of rules in that group and set both the fields we will only be able to set one to the limit. Should i open an issue in prometheus for support of two fields ?

I think that's fine; we can log a warning if the fields have different values, and use the smaller of the two.

Opening an issue on Prometheus sounds like a good idea 👍

**Comment by vardhaman-surana** [comment]
> > hey @dannykopping i started working on it and it seems two fields can not be added till those two fields are supported in prometheus. Since **limit** is configured at rule group level and it controls both the alerting and recording rules inside that rule group. So even if we expose two fields and a user has both kinds of rules in that group and set both the fields we will only be able to set one to the limit. Should i open an issue in prometheus for support of two fields ?
> 
> I think that's fine; we can log a warning if the fields have different values, and use the smaller of the two.
> 
> Opening an issue on Prometheus sounds like a good idea +1

logging a warning does seem a good idea .. but looking at the first glance user would feel that he can control both recording and alerting rules with different params and will set both in case .. and even if we use the smaller one then that would be same as using one field.
It will just misguide the user in my opinion.
@dannykopping 
I will open an issue on prometheus. And till that gets accepted can we merge this one ? 
so that at least users will have this limit option and it is clearly stated in loki docs that only prometheus compatible rules are supported so i guess adding this limit thing should be ok.
Please let me know what do you think should be the approach forward.

**Comment by dannykopping** [comment]
> so that at least users will have this limit option and it is clearly stated in loki docs that only prometheus compatible rules are supported so i guess adding this limit thing should be ok.

As a point of clarification, we _only_ support Prometheus-compatible rules.

> I will open an issue on prometheus. And till that gets accepted can we merge this one ?
If you add documentation which explains the difference between the behaviour, then yes I think we can.

**Comment by vardhaman-surana** [comment]
> > so that at least users will have this limit option and it is clearly stated in loki docs that only prometheus compatible rules are supported so i guess adding this limit thing should be ok.
> 
> As a point of clarification, we _only_ support Prometheus-compatible rules.
> 
> > I will open an issue on prometheus. And till that gets accepted can we merge this one ?
> 
> If you add documentation which explains the difference between the behaviour, then yes I think we can.

i can add the documentation. That's not a problem but essentially what will happen is only one field will be controlling both. so my suggestion is to have just one field, we can replace them when prometheus add support for two fields.

If you feel that adding two fields and having documentation is a good idea then i can make changes in this PR as well. Please let me know if i should go ahead. or the current limit thing is ok.

**Comment by dannykopping** [comment]
Sorry, I wasn't clear before.

Yes, let's go with one `Limit` field + docs, and once Prometheus adds two separate fields (might be tricky with backwards-compatibility) then we can revisit this.

**Comment by vardhaman-surana** [comment]
> Sorry, I wasn't clear before.
> 
> Yes, let's go with one `Limit` field + docs, and once Prometheus adds two separate fields (might be tricky with backwards-compatibility) then we can revisit this.

@dannykopping i have added docs for limit field. Please review the PR and let me know if any other changes are required. Thanks

**Comment by vardhaman-surana** [comment]
it seems there is a flaky test. also happened in the build https://drone.grafana.net/grafana/loki/26050/3/6 
can you please re run the build for the PR @dannykopping ?

**Comment by dannykopping** [comment]
Please remember to not force push!

File: docs/sources/alert/_index.md [hunk 1]
```diff
@@ -106,6 +106,40 @@ This query (`expr`) will be executed every 1 minute (`interval`), the result of
 name we have defined (`record`). This metric named `nginx:requests:rate1m` can now be sent to Prometheus, where it will be stored
 just like any other metric.
 
+
+### Limiting Alerts AND Series
```
Comment (dannykopping) [inline docs/sources/alert/_index.md:110]:
```suggestion
### Limiting Alerts and Recording Rule Samples
```

File: docs/sources/alert/_index.md [hunk 2]
```diff
@@ -106,6 +106,40 @@ This query (`expr`) will be executed every 1 minute (`interval`), the result of
...
+
+Like Prometheus, A limit for alerts produced by alerting rules and series produced by recording rules can be configured per-group in Loki also. When the limit is exceeded, all series produced by the rule are discarded, and if it's an alerting rule, all alerts for the rule, active, pending, or inactive, are cleared as well. Default value for limit is **0** i.e. no limit.
```
Comment (dannykopping) [inline docs/sources/alert/_index.md:112]: > Like Prometheus

Please link to Prometheus' documentation.

Also, I think we should describe _why_ someone should use this as well as simply _how_.

File: docs/sources/alert/_index.md (hunk 1 above)
Comment (vardhaman-surana) [reply docs/sources/alert/_index.md]: done

File: docs/sources/alert/_index.md (hunk 2 above)
Comment (vardhaman-surana) [reply docs/sources/alert/_index.md]: tried to explain a bit more here. please check.

**Review by dannykopping — APPROVED** [review]
LGTM, thanks very much for your patience @vardhaman-surana!

**Comment by vardhaman-surana** [comment]
@JStickler please review this.
Thanks.

**Review by JStickler — COMMENTED** [review]
[docs team]

File: docs/sources/alert/_index.md [hunk 3]
```diff
@@ -106,6 +106,40 @@ This query (`expr`) will be executed every 1 minute (`interval`), the result of
 name we have defined (`record`). This metric named `nginx:requests:rate1m` can now be sent to Prometheus, where it will be stored
 just like any other metric.
 
+
+### Limiting Alerts and Recording Rule Samples
+
+Like [Prometheus](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#limiting-alerts-and-series), A limit for alerts produced by alerting rules and samples produced by recording rules can be configured per-group in Loki also. This limit can be used to prevent a large number of alerts or recording samples from being generated due to a faulty rule. When the limit is exceeded, all recording samples produced by the rule are discarded, and if it's an alerting rule, all alerts for the rule, active, pending, or inactive, are cleared as well. The event will be recorded as an error in the evaluation, and the rule health will be set to `err`. Default value for limit is **0** i.e. no limit.
+
+#### Example
+
```
[the remark this answers is not in the document]
Comment (vardhaman-surana) [reply docs/sources/alert/_index.md:116]: done

File: docs/sources/alert/_index.md [hunk 4]
```diff
@@ -106,6 +106,41 @@ This query (`expr`) will be executed every 1 minute (`interval`), the result of
 name we have defined (`record`). This metric named `nginx:requests:rate1m` can now be sent to Prometheus, where it will be stored
 just like any other metric.
 
+
+### Limiting Alerts and Recording Rule Samples
+
+Like [Prometheus](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#limiting-alerts-and-series), you can configure a limit for alerts produced by alerting rules and samples produced by recording rules. This limit can be configured per-group. Using limits can prevent a faulty rule from generating a large number of alerts or recording samples. When the limit is exceeded, all recording samples produced by the rule are discarded, and if it is an alerting rule, all alerts for the rule, active, pending, or inactive, are cleared. The event will be recorded as an error in the evaluation, and the rule health will be set to `err`. The default value for limit is `0` meaning no limit.
+
+Here is an example of a rule group along with its limit configured.
+#### Example
```
Comment (JStickler) [inline docs/sources/alert/_index.md:115]:
```suggestion
#### Example

Here is an example of a rule group along with its limit configured.

```

**Review by JStickler — APPROVED** [review]
[docs team]  One small suggestion, but otherwise the docs LGTM.

**Comment by vardhaman-surana** [comment]
Thanks for the reviews @dannykopping @JStickler. This PR is now ready to merge

**Comment by dannykopping** [comment]
Thanks very much for your contribution @vardhaman-surana!

**Pull request merged**